### PR TITLE
Avatar demo

### PR DESCRIPTION
**Pull request by liPatrick**
Avatar demos with the following providers: HeyGen, DID, Microsoft, and Yepic. 
located at /avatar

### CODE REVIEW

File: src/app/avatar/heygen/helper.tsx [hunk 1]
```diff
@@ -0,0 +1,95 @@
+
+// new session
+export async function newSession(quality: any, avatar_name: any, voice_id: any) {
```
Comment (juberti) [inline src/app/avatar/heygen/helper.tsx]: maybe add a `doPost` helper so you don't have so much boilerplate here

File: src/app/avatar/api/heygen/newSession/route.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
+export async function POST(req: Request): Promise<Response> {
+  try {
```
Comment (juberti) [inline src/app/avatar/api/heygen/newSession/route.tsx]: I think uncaught exceptions are automatically turned into 500s.

File: src/app/avatar/api/heygen/newSession/route.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    const response = await fetch(`${SERVER_URL}/v1/streaming.new`, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+        'X-Api-Key': apiKey,
+      },
+      body: JSON.stringify({ quality, avatar_name, voice: { voice_id } }),
+    });
+
+    if (response.status === 500) {
```
Comment (juberti) [inline src/app/avatar/api/heygen/newSession/route.tsx]: can you just pass back the response object directly (or write some helper than converts it as needed?)

File: src/app/avatar/api/heygen/newSession/route.tsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
+export async function POST(req: Request): Promise<Response> {
+  try {
+    console.log('POST request received');
+    const body = await req.json();
+    const { quality, avatar_name, voice_id } = body;
+    
+    const SERVER_URL = process.env.HEYGEN_SERVER_URL || '';
```
Comment (juberti) [inline src/app/avatar/api/heygen/newSession/route.tsx]: maybe add some sort of doPost helper that minimizes the amount of boilerplate here

Comment (liPatrick) [reply src/app/avatar/api/heygen/newSession/route.tsx]: ah yes! much cleaner now

File: src/app/avatar/heygen/index.tsx [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+      mediaElementRef.current.onloadedmetadata = () => {
+        // Handle ready to play logic
+      };
+    }
+  }, [mediaElementRef]);
+
+  return (
+    <div className="flex flex-col gap-2">
+      <div>HeyGen Demo</div>
+      <p>1. Click connect.</p>
```
Comment (juberti) [inline src/app/avatar/heygen/index.tsx]: Suggest just having Start and having that do the connect automatically.

File: src/app/avatar/heygen/index.tsx [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+      <p>2. Click start. After a few seconds, you should see the avatar rendered.</p>
+      <p>3. Click repeat to have the avatar speak the text above.</p>
```
Comment (juberti) [inline src/app/avatar/heygen/index.tsx]: Change this to Speak, "repeat" is kind of confusing

File: src/app/avatar/heygen/index.tsx [hunk 7]
```diff
@@ -0,0 +1,174 @@
+'use client'
+import React, { useState, useEffect, useRef, useCallback } from 'react';
+import { handleICE, newSession, repeat, startSession, stopSession } from './helper';
+
+// Interface for session info
+interface SessionInfo {
+  session_id: string;
```
Comment (juberti) [inline src/app/avatar/heygen/index.tsx:7]: I think this is really the only data you need to store, not the other two fields.

File: src/app/avatar/heygen/index.tsx [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+      setMediaCanPlay(false);
+
+      console.log('Connection closed successfully');
+    } catch (err) {
+      console.error('Failed to close the connection:', err);
+      console.log('Failed to close the connection');
+    }
+  }, [sessionInfo]);
+  // useEffect for DOM element references
+  useEffect(() => {
```
Comment (juberti) [inline src/app/avatar/heygen/index.tsx]: Does this serve any purpose?

File: src/app/avatar/heygen/index.tsx [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+  // References to media elements
+  const mediaElementRef = useRef<HTMLVideoElement>(null);
+  const canvasElementRef = useRef<HTMLCanvasElement>(null); 
+
+  // Event handlers (with type annotations)
+  const createNewSession = async () => {
+    console.log('Creating new session... please wait');
+
+    // Simplified: Replace these with your actual logic to get these values
+    const avatar = ''; // Example, replace with actual
```
Comment (juberti) [inline src/app/avatar/heygen/index.tsx]: Did you mean to fill in some values here?

File: src/app/avatar/heygen/index.tsx [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+  // State variables
+  const [status, setStatus] = useState<string>('');
+  const [sessionInfo, setSessionInfo] = useState<SessionInfo | null>(null);
+  const [talkingToAI, setTalkingToAI] = useState(false);
+  const peerConnectionRef = useRef<RTCPeerConnection | null>(null);
+  const [mediaCanPlay, setMediaCanPlay] = useState<boolean>(false);
+
+  // References to media elements
+  const mediaElementRef = useRef<HTMLVideoElement>(null);
+  const canvasElementRef = useRef<HTMLCanvasElement>(null);
```
Comment (juberti) [inline src/app/avatar/heygen/index.tsx]: Is this used?

File: src/app/avatar/heygen/index.tsx [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+
+export function HeyGenPage({text}:{text: string}) {
+  // State variables
+  const [status, setStatus] = useState<string>('');
+  const [sessionInfo, setSessionInfo] = useState<SessionInfo | null>(null);
+  const [talkingToAI, setTalkingToAI] = useState(false);
+  const peerConnectionRef = useRef<RTCPeerConnection | null>(null);
+  const [mediaCanPlay, setMediaCanPlay] = useState<boolean>(false);
```
Comment (juberti) [inline src/app/avatar/heygen/index.tsx]: Is this used?

File: src/app/avatar/heygen/index.tsx [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+  sdp: RTCSessionDescriptionInit;
+  ice_servers2: RTCIceServer[];
+}
+
+
+export function HeyGenPage({text}:{text: string}) {
+  // State variables
+  const [status, setStatus] = useState<string>('');
+  const [sessionInfo, setSessionInfo] = useState<SessionInfo | null>(null);
+  const [talkingToAI, setTalkingToAI] = useState(false);
```
Comment (juberti) [inline src/app/avatar/heygen/index.tsx]: Is this used?

File: src/app/avatar/heygen/index.tsx [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+interface SessionInfo {
+  session_id: string;
+  sdp: RTCSessionDescriptionInit;
+  ice_servers2: RTCIceServer[];
+}
+
+
+export function HeyGenPage({text}:{text: string}) {
+  // State variables
+  const [status, setStatus] = useState<string>('');
```
Comment (juberti) [inline src/app/avatar/heygen/index.tsx]: Is this used?

File: src/app/avatar/heygen/index.tsx [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+  const [sessionInfo, setSessionInfo] = useState<SessionInfo | null>(null);
+  const [talkingToAI, setTalkingToAI] = useState(false);
+  const peerConnectionRef = useRef<RTCPeerConnection | null>(null);
```
Comment (juberti) [inline src/app/avatar/heygen/index.tsx:14]: Might be worth breaking this out into its own class - I bet the other implementations will need something similar

File: src/app/avatar/heygen/index.tsx (hunk 9 above)
Comment (liPatrick) [reply src/app/avatar/heygen/index.tsx]: I think we can leave them blank for now

File: src/app/avatar/heygen/index.tsx (hunk 14 above)
Comment (liPatrick) [reply src/app/avatar/heygen/index.tsx:14]: I think I'll break it out after I've connected a few more providers, so its clear to me where the overlap is.

File: src/app/avatar/api/did/helper.tsx [hunk 15]
```diff
@@ -0,0 +1,22 @@
+
+
+export async function doPost(url: string, data: any) {
```
Comment (juberti) [inline src/app/avatar/api/did/helper.tsx:3]: `url` here and below is more of a path, right?

**Review by juberti — APPROVED** [review]
This is a good start, feel free to land this once you've gotten things up and running

**Pull request merged**